File: hooks/block-platform-specific-scripts.py
#!/usr/bin/env python3
"""PreToolUse hook (Edit|Write): block platform-specific script creation.

.sh/.bash/.zsh/.csh/.tcsh/.fish/.ksh → Windows非対応のためブロック
.ps1/.psm1/.psd1/.bat/.cmd/.vbs/.vbe/.wsf/.wsh → Linux非対応のためブロック
いずれも .py への変更を促す。
"""

import json
import os
import sys

UNIX_ONLY_EXTENSIONS = {".sh", ".bash", ".zsh", ".csh", ".tcsh", ".fish", ".ksh"}
WINDOWS_ONLY_EXTENSIONS = {".ps1", ".psm1", ".psd1", ".bat", ".cmd", ".vbs", ".vbe", ".wsf", ".wsh"}


def main() -> None:
    data = json.load(sys.stdin)
    tool_input = data.get("tool_input", {})
    file_path = tool_input.get("file_path", "") or tool_input.get("path", "")

    if not file_path:
        return

    ext = os.path.splitext(file_path)[1].lower()

    if ext in UNIX_ONLY_EXTENSIONS:
        json.dump(
            {
                "decision": "block",
                "reason": f"それはWindowsでは使えません。.pyにしてください。（{ext}）",
            },
            sys.stdout,
        )
    elif ext in WINDOWS_ONLY_EXTENSIONS:
        json.dump(
            {
                "decision": "block",
                "reason": f"それはLinuxでは使えません。.pyにしてください。（{ext}）",
            },
            sys.stdout,
        )


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/block_platform_specific_scripts.rs
use regex::Regex;

use crate::input::HookInput;
use crate::output::{block, pass};

const UNIX_ONLY: &[&str] = &[".sh", ".bash", ".zsh", ".csh", ".tcsh", ".fish", ".ksh"];
const WINDOWS_ONLY: &[&str] = &[
    ".ps1", ".psm1", ".psd1", ".bat", ".cmd", ".vbs", ".vbe", ".wsf", ".wsh",
];

fn classify(ext: &str) -> Option<&'static str> {
    if UNIX_ONLY.contains(&ext) {
        return Some("Unix");
    }
    if WINDOWS_ONLY.contains(&ext) {
        return Some("Windows");
    }
    None
}

fn ext_reason(ext: &str) -> Option<String> {
    match classify(ext)? {
        "Unix" => Some(format!(
            "それはWindowsでは使えません。.pyにしてください。（{ext}）"
        )),
        _ => Some(format!(
            "それはLinuxでは使えません。.pyにしてください。（{ext}）"
        )),
    }
}

pub fn run(input: &HookInput) {
    let file_path = input.tool_input.file_path_resolved();
    if !file_path.is_empty() {
        let ext = std::path::Path::new(file_path)
            .extension()
            .and_then(|e| e.to_str())
            .unwrap_or("");
        let ext_lower = format!(".{ext}").to_lowercase();
        if let Some(reason) = ext_reason(&ext_lower) {
            block(&reason);
            return;
        }
    }

    let command = &input.tool_input.command;
    if !command.is_empty() {
        let redirect_re = Regex::new(r#">{1,2}\s*["']?(\S+?)["']?\s*$"#).unwrap();
        let touch_re = Regex::new(r#"\btouch\s+(?:-\S+\s+)*["']?(\S+?)["']?\s*$"#).unwrap();
        let tee_re = Regex::new(r#"\btee\s+(?:-\S+\s+)*["']?(\S+?)["']?\s*$"#).unwrap();

        for re in [&redirect_re, &touch_re, &tee_re] {
            if let Some(caps) = re.captures(command)
                && let Some(m) = caps.get(1)
            {
                let filepath = m.as_str();
                let ext = std::path::Path::new(filepath)
                    .extension()
                    .and_then(|e| e.to_str())
                    .unwrap_or("");
                let ext_lower = format!(".{ext}").to_lowercase();
                if let Some(reason) = ext_reason(&ext_lower) {
                    block(&reason);
                    return;
                }
            }
        }
    }

    pass();
}
